File: Graphs/Easy/dfs_of_graph.py
#User function Template for python3
# https://practice.geeksforgeeks.org/problems/depth-first-traversal-for-a-graph/1#
class Solution:
    arr = []
    visited = []
    #Function to return a list containing the DFS traversal of the graph.
    def dfsOfGraph(self, V, adj):
        self.arr = []
        self.visited = [False] * V
        self.dfs(adj, 0)
        
        return self.arr
        
    def dfs(self, adj, idx):
        if(self.visited[idx]):
            return
            
        self.visited[idx] = True
        self.arr.append(idx)
        
        for i in adj[idx]:
            if not self.visited[i]:
                self.dfs(adj, i)
        

#{ 
#  Driver Code Starts
if __name__ == '__main__':
    T=int(input())
    while T>0:
        V,E=map(int,input().split())
        adj=[[] for i in range(V+1)]
        for i in range(E):
            u,v=map(int,input().split())
            adj[u].append(v)
            adj[v].append(u)
        ob=Solution()
        ans=ob.dfsOfGraph(V,adj)
        for i in range(len(ans)):
            print(ans[i],end=" ")
        print()
        T-=1
# } Driver Code Ends

File: Graphs/Easy/dfs_of_graph.cpp
// https://practice.geeksforgeeks.org/problems/depth-first-traversal-for-a-graph/1#
// { Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

 // } Driver Code Ends
class Solution {
  public:
    // Function to return a list containing the DFS traversal of the graph.
    vector<int> dfsOfGraph(int V, vector<int> adj[]) {
        vector<bool> visited(V, false);
        vector<int> arr;
        
        dfs(adj, visited, arr, 0);
        
        return arr;
    }
    
    void dfs(vector<int> adj[], vector<bool>& visited, vector<int>& arr, int i)
    {
        if(visited[i]) return;
        visited[i] = true;
        arr.push_back(i);
        for(auto idx : adj[i])
        {
            if(!visited[idx])
                dfs(adj, visited, arr, idx);
        }
    }
};

// { Driver Code Starts.
int main() {
    int tc;
    cin >> tc;
    while (tc--) {
        int V, E;
        cin >> V >> E;

        vector<int> adj[V];

        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
            adj[v].push_back(u);
        }
        // string s1;
        // cin>>s1;
        Solution obj;
        vector<int> ans = obj.dfsOfGraph(V, adj);
        for (int i = 0; i < ans.size(); i++) {
            cout << ans[i] << " ";
        }
        cout << endl;
    }
    return 0;
}  // } Driver Code Ends